Add public API with a request object for Select/Update/Upstream

oleg-jukovec opened the pull request:
This patch provides request types for part of space operations: Select, Update
and Upstream. It allows to create requests step by step. The main idea here
is too provide more extensible approach to create requests.

Part of #126

ligurio commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 ### Added
 
 - Coveralls support (#149)
+- Add public API with a request object for Select/Update/Upstream (#126)
```
We usually add changelog entries to the end of the list.

oleg-jukovec replied:
Oh, it looks like an order by an issue number. Thank you, I got it.

ligurio commented on client_tools.go:
```diff
@@ -67,6 +67,80 @@ func (o Op) EncodeMsgpack(enc *msgpack.Encoder) error {
 	return enc.Encode(o.Arg)
 }
 
+// We don't want to make it public.
```
I think comment is not needed here. All constants with lower first letter are not public.

oleg-jukovec replied:
Okay, fixed.

vr009 commented on future.go:
```diff
@@ -0,0 +1,126 @@
+package tarantool
```
It seems, that these changes are not related to the issue or related indirectly, maybe it is worth to move it in a separate commit.

oleg-jukovec replied:
Good idea. Done.

vr009 commented on tarantool_test.go:
```diff
@@ -92,6 +92,37 @@ func BenchmarkClientSerial(b *testing.B) {
 	}
 }
 
+func BenchmarkClientSerialRequestObject(b *testing.B) {
```
Probably it would be helpful to add a Benchmark with the parallel load, where you call b.RunParallel inside. That would show the full difference between the existing API and the proposed one. Feel free to ignore, I have already implemented it in scope of my patch, which is over your one. On the other hand my patch could be discarded and we will lose those Benchmarks.

oleg-jukovec replied:
I don't think that request objects will somehow critically affect parallel execution in comparison with single-threaded. The benchmark will look better in the issue with contexts.

vr009 commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 - queue-utube handling (#85)
 - Master discovery (#113)
 - SQL support (#62)
+- Add public API with a request object for Select/Update/Upstream (#126)
```
There is a word "Added" above, seems the "Add" here is unnecessary.

oleg-jukovec replied:
Thank you, fixed.

vr009 commented on const.go:
```diff
@@ -1,19 +1,19 @@
-	ExecuteRequest   = 11
-	PingRequest      = 64
-	SubscribeRequest = 66
+	SelectRequestCode    = 1
```
This change seems not worth to be moved in a separate commit, but I would mentioned this change in the commit message.

oleg-jukovec replied:
Done. I've also added this to CHANGELOG.md because that constants are available from a client code.

vr009 reviewed (approved):
The patch looks good to me. Please, consider my comments. 

Also there is a kind of a "rule" to add some prefix word in the commit header, like "api: add a request object for select/update/upsert". See `CONTRIBUTING.md` for details.

What about the implementation of request object, and it's filling with chaining methods, I like this approach. It is more comfortable, to use the constructor and other methods instead of filling the structure manually and passing it.

oleg-jukovec commented:
I've updated the pull request. Changes:

* add request object API for all queries
* add `Do`/`DoTyped`/`DoAsync` to ConnectionPool and ConnectionMulti types
* fix some comments

In additional, we can use Request objects in the internal implementation instead of closures:
https://github.com/tarantool/go-tarantool/blob/d44ffa0cda16579bf8ec41534067b29762f1326d/request.go#L272-L275

It will simplify the `Request` interface:
```go
type Request interface {
	Code() int32
	BodyFunc(resolver SchemaResolver) (func(enc *msgpack.Encoder) error, error)
}
```
There will be no need to return a closure from `BodyFunc` and it can be replaced by just `Body(enc *msgpack.Encoder)`. I like this change, but I would like to get some feedback first.

Totktonada commented on request.go:
```diff
@@ -508,3 +520,212 @@ func encodeSQLBind(enc *msgpack.Encoder, from interface{}) error {
+	return req
+}
+
+// Limit sets the limit for the select request.
+// Note: default value is 0.
```
AFAIU we send 2^32-1 as the limit to the protocol if the limit is not sent on the client side. How it works here? Zero is interpreted as the alias for 2^32-1? A user can't pass zero to verify that the space is exists and accessible?

oleg-jukovec replied:
Thank you, replaced to 0xFFFFFFFF by default.

Totktonada commented on request.go:
```diff
@@ -508,3 +520,212 @@ func encodeSQLBind(enc *msgpack.Encoder, from interface{}) error {
+	return req
+}
+
+// Iterator set the iterator for the select request.
+// Note: default value is IterEq.
```
Should not it be `IterEQ` when the key is provided and is not empty and `IterAll` otherwise?

I just seen the [similar logic](https://github.com/tarantool/tarantool/blob/ede831d317c3a5ba02f202a7daf387742ddbfeb6/src/box/lua/schema.lua#L1892-L1893) in tarantool's Lua APi implementation and wonder, whether something like this should be there.

oleg-jukovec replied:
Good idea, done.

Totktonada replied:
What if the key is set to `null` or an empty array? I don't know, to be honest, whether it makes any difference in responses. If not, feel free to choose any solution.

oleg-jukovec replied:
It seems to me that we should not try to validate the input parameters here. As we have not tried to do this anywhere before (`Connection.Select` as example).

The user can still shoot himself in the foot no matter how hard we try. It will be clear not to validate the input parameters for all cases.

Totktonada replied:
Okay.

Totktonada commented on tarantool_test.go:
```diff
@@ -1549,6 +1580,113 @@ func TestClientNamed(t *testing.T) {
+		}
+	}
+
+	// Update without operations.
```
Is it legal? What we'll get in the response? Is it worth to tests responses (as I see, we only test response for a select request)?

oleg-jukovec replied:
I think it is legal until a tarantool instance does not return an error. It returns unmodified tuple.

I've added checks for responses.

Totktonada replied:
Okay. It seems, it is actually legal. I didn't know.

Totktonada commented on request.go:
```diff
@@ -729,3 +884,108 @@ func (req *UpsertRequest) BodyFunc(res SchemaResolver) (func(enc *msgpack.Encode
+	return req
+}
+
+// Args sets the args for the call request.
+// Note: default value is empty.
+func (req *CallRequest) Args(args interface{}) *CallRequest {
+	req.args = args
+	return req
+}
```
Shouldn't we provide something convenient for adding positional and named binding parameters?

CC @vr009.

Totktonada replied:
(It is about `ExecuteRequest`, I placed the comment to the wrong place.)

vr009 replied:
AFAIU you mean something like this:
```go 
ExecuteReqObj.Args([]interface{uuid.New(), "Value_A", 14})

...
...

ExecuteReqObj.PosArg(1, uuid.New()).PosArg(2, "Value_B").NamedArg("count", 15)
```
That looks to be useful when it is needed to change some field explicitly.

oleg-jukovec replied:
> That looks to be useful when it is needed to change some field explicitly.

For Execute, we can make special Request Objects, as examples:

https://gobuffalo.io/documentation/database/querying/
https://v1.gorm.io/docs/query.html
https://v1.gorm.io/docs/query.html#Scan
https://v1.gorm.io/docs/query.html#Group-amp-Having

The only thing is that it can be difficult to implement.

vr009 replied:
I'm ok with the current implementation, so I'm not insisting.

Totktonada replied:
I guess that nothing will prevent us to add more construction methods for binding parameter in a future. So I'll resolve the discussion. This PR is not where we plan to dig into SQL a lot.

Totktonada reviewed (approved):
I think it is the nice addition!

Looks okay for me at brief glance.

Totktonada commented:
> I like this change and I want to do this, but I would like to get some feedback first.

I don't see any downsides in the proposed refactoring.

oleg-jukovec commented:
I've updated the pull request. Changes:

- Commits with request objects (Select/Update/Upsert and Insert/.../Ping) have been squashed into one.
- Added a commit with refactoring. Request objects are used by an internal implementation instead of closures

DifferentialOrange commented on CHANGELOG.md:
```diff
@@ -12,9 +12,13 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Changed
 
+- IPROTO constants that identify requests renamed from `NameRequest` to
```
I think it's better to use something like `<Name>Request` because until I've seen the code changes I didn't understood that it meant a group of constants

Totktonada replied:
By the way, they're named `IPROTO_*` in tarantool sources.

oleg-jukovec replied:
> I think it's better to use something like `<Name>Request` because until I've seen the code changes I didn't understood that it meant a group of constants

It confused me too. Thank you, I updated it.

oleg-jukovec replied:
> By the way, they're named `IPROTO_*` in tarantool sources.

Thanks. I'll try to remember. Updated.

oleg-jukovec commented on connection.go:
```diff
@@ -468,18 +468,35 @@ func (conn *Connection) dial() (err error) {
-		requestCode: AuthRequest,
+func pack(h *smallWBuf, enc *msgpack.Encoder, reqid uint32, req Request, res SchemaResolver) (err error) {
+	hl := h.Len()
+	h.Write([]byte{
```
writeAuthRequest:
https://github.com/tarantool/go-tarantool/blob/1a1fe7b4e2ea77aa316ed4006aa445e512812c3c/connection.go#L471

pack:
https://github.com/tarantool/go-tarantool/blob/1a1fe7b4e2ea77aa316ed4006aa445e512812c3c/future.go#L24

DifferentialOrange replied:
Thank you, `pack` is what I was searching for

oleg-jukovec commented on connection.go:
```diff
@@ -978,6 +988,32 @@ func (conn *Connection) nextRequestId() (requestId uint32) {
+//
+// An error is returned if the request was formed incorrectly, or failure to
+// communicate by the connection, or unable to decode the response.
+func (conn *Connection) Do(req Request) (*Response, error) {
```
It was Exec() or Execute(), but we already have Execute() for SQL. The [http module has Do](https://pkg.go.dev/net/http#Client.Do) in the same meaning, so I think it's ok.

DifferentialOrange reviewed (approved):
I couldn't say I dig too deep, but it seems LGTM

oleg-jukovec commented on request.go:
```diff
@@ -508,3 +535,465 @@ func encodeSQLBind(enc *msgpack.Encoder, from interface{}) error {
+
+// BodyFunc returns a function that can create an encoded body of
+// the ping request.
+func (req *PingRequest) BodyFunc(res SchemaResolver) (func(enc *msgpack.Encoder) error, error) {
```
It returns function with type `func(enc *msgpack.Encoder) error` or an error if unable to return the function.

oleg-jukovec commented on example_test.go:
```diff
@@ -126,6 +126,116 @@ func ExampleConnection_SelectAsync() {
+	}
+	fmt.Printf("response is %#v\n", resp.Data)
+	// Output:
+	// response is []interface {}{}
```
This is output for the first Upsert request:
https://github.com/tarantool/go-tarantool/blob/2d477f455c42add4df982767d8612b4bae59fc6e/example_test.go#L193-L202

Upserts responses are always empty:
https://github.com/tarantool/go-tarantool/blob/2d477f455c42add4df982767d8612b4bae59fc6e/request.go#L142-L143

DifferentialOrange replied:
Got it, thanks

Totktonada commented:
@oleg-jukovec Are you plan to finish any work here or it is time to proceed?

oleg-jukovec commented:
> @oleg-jukovec Are you plan to finish any work here or it is time to proceed?

I plan to merge it after #171 .
